alonzo/formal-spec/transactions.tex

```diff
@@ -138,33 +119,33 @@ \section{Transactions}
        & (\VKey \mapsto \Sig) & \fun{txwitsVKey} & \text{VKey signatures}\\
        & \times ~(\ScriptHash \mapsto \Script)  & \fun{txscripts} & \text{All scripts}\\
        & \times~ \hldiff{(\DataHash \mapsto \Data)} & \hldiff{\fun{txdats}} & \text{All datum objects}\\
-       & \times ~\hldiff{(\RdmrPtr \mapsto \Data)}& \hldiff{\fun{txrdmrs}}& \text{Indexed redeemers}\\
+       & \times ~\hldiff{(\RdmrPtr \mapsto \Data \times \ExUnits)}& \hldiff{\fun{txrdmrs}}& \text{Redeemers/budget}\\
     \end{array}
   \end{equation*}
   %
   \begin{equation*}
     \begin{array}{r@{~~}l@{~~}l@{\qquad}l}
       \var{txbody} ~\in~ \TxBody ~=~
       & \powerset{\TxIn} & \fun{txinputs}& \text{Inputs}\\
```

Comment:
mention that these are specifically non-fee inputs (either in `\text{Non-fee inputs}` or in the prose)

Reply:
Do you mean the txinputs field? That's technically not the case: a transaction that doesn't have a non-native script doesn't need to use the txinputs_fee field at all, and can pay its fees as in Shelley.

Reply:
I see that instead requiring the fee-inputs cover the fee always, it is now required to be the case only when exunits are non-0,
`&~~      \wedge~ (\fun{exunits}~tx \neq (0, 0) \Rightarrow \var{balance} \geq \txfee{txb}) \\`

I dont see a check that the `balance \geq \txfee{txb}` when this isnt the case (eg Shelley Txs), but it should always be the case. At the time of translating a Tx from Shelley to Alonzo, all inputs should always be marked as for-fees.

What I was originally pointing out, though, is that it is not clear from the name that these two sets of inputs are disjoint sets (which we should also check, i believe). So, "for-fee" and "not for-fee" would make sense.

Reply:
Also, the explanation of `feesOK` has
    \item the fee-marked inputs do not have attached data, which implies that they
      cannot be validated by non-native scripts;
      
Which isnt really what the first check in the function is saying? What i think it is saying is that the for-fees inputs are not non-native scripts. Which is more correct than bringing attached datums into the explanation. I would restate that item as
"\item the fee-marked inputs are not locked by non-native scripts;"

and add, as a comment below, that we dont actually disallow attaching datums to Key Hash outputs or Native script outputs, but doing this will just cost the tx author more.

Reply:
The intent here is actually the opposite: a Shelley transaction should be interpreted as having no fee inputs. Basically, for a transaction that doesn't use non-native scripts the `feesOK` check should just reduce to the old check from Shelley. This means that fees can also be covered for example by things that aren't inputs.

However, I just realized that there is no `exunits` field anymore, so we should probably instead check that there are no redeemers. Also, good catch that the two input sets should be disjoint!

Reply:
You are right about feesOK!

And yes, I guess check that there are no redeemers. And also - the `minfee` calculation now has to add up all the `ExUnits` i guess?